design/httpproxy-route-sorting.md

```diff
@@ -0,0 +1,170 @@
+
+### Basic Case
+
+In the basic case a single `HTTPProxy` with the following routes:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example
+spec:
+  virtualhost:
+    fqdn: example.com
+  routes:
+    - conditions:
+      - prefix: /
+      - header:
+          exact: bar
+          name: x-experiment
+      services:
+        - name: b1
+          port: 400
+    - conditions:
+      - exact: /bar
+      services:
+        - name: b2
+          port: 400
+    - conditions:
+      - exact: /
+      services:
+        - name: b3
+          port: 400
+```
+
+Would result in the following:
+
+| Request                          | With Sorting  | Without Sorting |
+| -------------                    | ------------- | --------------- |
+| GET -H "x-experiment: bar" /bar  | b2            | b1              |
+| GET /bar | Content Cell          | b2            | b2              |
+| GET /                            | b3            | b3              |
+
+
+### Route Matching with Include Statements
+
+Include statements are resolved in a depth first fashion.
+
+```yaml
+# httpproxy-inclusion-samenamespace.yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example-include
+  namespace: default
+spec:
+  virtualhost:
+    fqdn: example.com
+  includes:
+  - name: service2
+    namespace: default
+  routes:
+    - conditions:
+      - prefix: /bar
+      services:
+        - name: s1
+          port: 80
+      - prefix: /
+      services:
+        - name: s2
+          port: 80
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: service2
+  namespace: default
+spec:
+  routes:
+    - conditions:
+      - prefix: /blog
+      services:
+        - name: s3
+          port: 80
+```
+
+Will result in the following traffic table in Envoy:
+
+```yaml
+virtual_hosts:
+- name: example
+  domains: ["example.com"]
+  routes:
+  - match:
+      prefix: /blog
+    route:
+      cluster: s3
+  - match:
+      prefix: /bar
+    route:
+      cluster: s1
+  - match:
+      prefix: /
+    route:
+      cluster: s2
+
+```
+
+
+### Route Matching for Objects managed by both an HTTPProxy and Ingress (or Gateway)
+
+Since `Gateway` and `Ingress` have routing conformance which are defined outside of the repository any `virtualHost` that is managed by both `CRDs` will follow the `Gateway`/`Ingress` sorting regardless of
+the configuration.
+
+## Alternatives Considered
```

Comment:
Another alternative is to just offer guidance via documentation about what users configuring contour should do to realize certain routing schemes.

e.g. add guidance from this comment: https://github.com/projectcontour/contour/issues/5773#issuecomment-1779332101

this may take the form of generating additional routing rules with specific matches to get the precedence behavior desired

pros:
- no changes to contour
- contour routing stays similar to the ingress/gateway api standard for all users so migration is not an issue

cons:
- could lead to proliferation of special case rules that are hard to reason about for humans
- no guarantee the exact routing specificity can be acheived

Reply:
would be nice to include this in the document at least for discussion purposes/posterity

Reply:
yeah fair comment, added